src/core/service.c

```diff
@@ -2758,6 +2761,63 @@ static void service_run_next_main(Service *s) {
         (void) service_set_main_pidref(s, TAKE_PIDREF(pidref), &s->main_command->exec_status.start_timestamp);
 }
 
+static bool service_needs_reload_extensions(Unit *u) {
+        Service *s = ASSERT_PTR(SERVICE(u));
+        int r;
+
+        /* Only support this for notify-reload service types. */
+        if (s->type != SERVICE_NOTIFY_RELOAD)
+                return false;
+
+        if (s->exec_context.n_extension_images == 0 && strv_isempty(s->exec_context.extension_directories))
+                return false;
+
+        r = exec_context_has_vpick_extensions(&s->exec_context);
+        if (r < 0) {
+                log_unit_warning_errno(u, r, "Failed to determine if service needs reloading extensions, assuming false: %m");
+                return false;
+        }
+        return r == 1;
+}
+
+static int service_reload_extensions(Unit *u) {
+        /* TODO: do this asynchronously */
+        _cleanup_free_ char *propagate_dir = NULL;
+
+        Service *s = ASSERT_PTR(SERVICE(u));
+
+        /* TODO: Add support for user services, which can use
+         * ExtensionDirectories= + notify-reload. For now, skip for user
+         * services. */
+        if (u->manager->runtime_scope != RUNTIME_SCOPE_SYSTEM) {
+                log_unit_info(UNIT(s), "Extension refresh is not currently supported for user service.");
```

Comment:
We don't really log about "currently". Aside from that, let's not log about this loud and every time a user service with extensions gets reloaded. Move this check to `..._should_reload_``` and use something like `log_debug("Not reloading extensions for user services")`, please